File: internal/runner/events.go
// Package runner implements the Docker container runner for agent execution.
package runner

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/kalvis/mesh/internal/model"
)

// HumanizeEvent returns a short human-readable summary of an agent event.
// This is observability-only output and must not affect orchestrator logic.
func HumanizeEvent(ev model.AgentEvent) string {
	switch ev.Event {
	case "session_started":
		return "Session started"
	case "turn_started":
		if ev.Turn > 0 {
			return fmt.Sprintf("Turn %d started", ev.Turn)
		}
		return "Turn started"
	case "turn_completed":
		if ev.Turn > 0 {
			return fmt.Sprintf("Turn %d completed", ev.Turn)
		}
		return "Turn completed"
	case "completed":
		if ev.TurnsUsed > 0 {
			return fmt.Sprintf("Session completed (%d turns)", ev.TurnsUsed)
		}
		return "Session completed"
	case "notification":
		if ev.Message != "" {
			msg := ev.Message
			if len(msg) > 80 {
				msg = msg[:77] + "..."
			}
			return fmt.Sprintf("Agent: %s", msg)
		}
		return "Agent notification"
	case "usage":
		return fmt.Sprintf("Usage: %d in / %d out tokens", ev.InputTokens, ev.OutputTokens)
	case "error":
		if ev.Message != "" {
			msg := ev.Message
			if len(msg) > 80 {
				msg = msg[:77] + "..."
			}
			return fmt.Sprintf("Error: %s", msg)
		}
		return "Agent error"
	case "malformed":
		return "Malformed agent output"
	default:
		return fmt.Sprintf("Event: %s", ev.Event)
	}
}

// ParseEvent parses a single JSON line from the agent's stdout into an AgentEvent.
// Returns an error if the line is empty, not valid JSON, or missing the required "event" field.
func ParseEvent(line string) (model.AgentEvent, error) {
	trimmed := strings.TrimSpace(line)
	if trimmed == "" {
		return model.AgentEvent{}, fmt.Errorf("empty line")
	}

	var ev model.AgentEvent
	if err := json.Unmarshal([]byte(trimmed), &ev); err != nil {
		return model.AgentEvent{}, fmt.Errorf("invalid JSON: %w", err)
	}

	if ev.Event == "" {
		return model.AgentEvent{}, fmt.Errorf("missing required field: event")
	}

	return ev, nil
}

File: agent/src/mesh_agent/events.py
"""JSON line event emitter for stdout communication with Go orchestrator."""

from __future__ import annotations

import json
import sys
from datetime import datetime, timezone


class EventEmitter:
    """Emits JSON line events to stdout."""

    def __init__(self, session_id: str) -> None:
        self.session_id = session_id

    def emit(self, event: str, **kwargs: object) -> None:
        """Emit a single JSON line event to stdout."""
        payload = {
            "event": event,
            "ts": datetime.now(timezone.utc).isoformat(),
            "session_id": self.session_id,
            **kwargs,
        }
        line = json.dumps(payload, default=str)
        sys.stdout.write(line + "\n")
        sys.stdout.flush()

    def session_started(self) -> None:
        self.emit("session_started")

    def turn_started(self, turn: int) -> None:
        self.emit("turn_started", turn=turn)

    def turn_completed(
        self, turn: int, message: str = "", input_tokens: int = 0, output_tokens: int = 0
    ) -> None:
        self.emit(
            "turn_completed",
            turn=turn,
            message=message,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
        )

    def completed(
        self, turns_used: int, input_tokens: int = 0, output_tokens: int = 0, total_tokens: int = 0
    ) -> None:
        self.emit(
            "completed",
            turns_used=turns_used,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            total_tokens=total_tokens,
        )

    def notification(self, message: str) -> None:
        """Emit a human-readable status update."""
        self.emit("notification", message=message)

    def usage(self, input_tokens: int, output_tokens: int, total_tokens: int) -> None:
        """Emit a token consumption update."""
        self.emit(
            "usage",
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            total_tokens=total_tokens,
        )

    def error(self, code: str, message: str) -> None:
        self.emit("error", code=code, message=message)
